apollo-router/src/plugins/response_cache/plugin.rs

```diff
@@ -1345,44 +1341,59 @@ fn get_invalidation_root_keys_from_schema(
-                &supergraph_schema,
-                executable_document,
-                variable_values,
-                &mut errors,
-                Default::default(),
-                field_def,
-                field,
-            )
-            .map_err(|_| FetchError::MalformedRequest {
-                reason: format!("cannot argument values for root fields {:?}", field.name),
-            })?;
-
-            if !errors.is_empty() {
-                return Err(FetchError::MalformedRequest {
-                    reason: format!(
-                        "cannot coerce argument values for root fields {:?}, errors: {errors:?}",
-                        field.name,
-                    ),
-                }
-                .into());
-            }
 
             let mut vars = IndexMap::default();
-            vars.insert("$args".to_string(), Value::Object(args));
-            cache_keys
-                .map(|ck| Ok(ck.interpolate(&vars).map(|(res, _)| res)?))
-                .collect::<Result<Vec<String>, anyhow::Error>>()
-        })
-        .collect::<Result<Vec<Vec<String>>, anyhow::Error>>()?;
+            vars.insert("$args".to_string(), Value::Object(info.arguments().clone()));
 
-    let invalidation_cache_keys: HashSet<String> = cache_keys.into_iter().flatten().collect();
+            for template in templates {
+                match template.interpolate(&vars) {
+                    Ok((key, _)) => {
+                        keys.insert(key);
+                    }
+                    Err(e) => {
+                        *result = Err(e.into());
+                        break;
+                    }
+                }
+            }
+            Ok(resolvers::ResolvedValue::SkipForPartialExecution)
+        }
+    }
 
-    Ok(invalidation_cache_keys)
+    let executable_document =
+        request
+            .executable_document
+            .as_ref()
+            .ok_or_else(|| FetchError::MalformedRequest {
+                reason: "cannot get the executable document for subgraph request".to_string(),
+            })?;
+    let root_query_type = supergraph_schema
+        .root_operation(apollo_compiler::ast::OperationType::Query)
+        .ok_or_else(|| FetchError::MalformedRequest {
+            reason: "cannot get the root operation from supergraph schema".to_string(),
+        })?;
+    let query_object_type = supergraph_schema
+        .get_object(root_query_type.as_str())
+        .ok_or_else(|| FetchError::MalformedRequest {
+            reason: "cannot get the root query type from supergraph schema".to_string(),
+        })?;
+    let root = Root {
+        subgraph_name: &request.subgraph_name,
+        subgraph_enums,
+        query_object_type,
+        result: RefCell::new(Ok(HashSet::new())),
+    };
+    let subgraph_request = request.subgraph_request.body();
+    // FIXME: in principle we should use the subgraph schema here.
+    // Maybe this is good enough as far as finding root fields is concerned?
```

Comment:
I think it's fine here though i'm not 1000% sure; it's also what the previous implementation did though, so not a regression if it is indeed subtly wrong.